### PR TITLE
feat: add online data mixing plugin

### PR DESCRIPTION
**Pull request by kmehant**
Details provided in https://github.com/foundation-model-stack/fms-acceleration/pull/152

### CODE REVIEW

File: tuning/data/data_processors.py [hunk 1]
```diff
@@ -452,8 +452,22 @@ def split_dataset(
...
+            if eval_split in raw:
+                eval_datasets_dict[d.name] = raw[eval_split]
+        print("train_datasets_dict", train_datasets_dict)
+        print("eval_datasets_dict", eval_datasets_dict)
+        return train_datasets_dict, eval_datasets_dict
+
     def _process_dataset_configs(
-        self, dataset_configs: List[DataSetConfig]
+        self, dataset_configs: List[DataSetConfig], odm_config=None
     ) -> Union[Dataset, IterableDataset]:
```
[the remark this answers is not in the document]
Comment (kmehant) [reply tuning/data/data_processors.py:469]: @romitjain May I know which annotation you are talking about? The return type of `_process_dataset_configs `? In our case it will be `IterableDataset ` isn't it?

**Comment by kmehant** [comment]
NOTE: format/lint error should go once we have fms_acceleration_odm package available.

**Review by ashokponkumar — DISMISSED** [review]

File: tests/artifacts/predefined_data_configs/multiple_datasets_with_odm.yaml [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+      update_interval: 1 # update every step
+      sampling_interval: 1 # sample category for every sample
+      reward_type: validation_loss # uses eval loss of each dataset as reward
+      gamma: 0.1 # MAB hyper-parameter
+      eta: 0.2 # MAB hyper-parameter
+datasets:
+  - name: dataset_1
+    split:
+      train: 0.8
+      validation: 0.2
```
Comment (dushyantbehl) [inline tests/artifacts/predefined_data_configs/multiple_datasets_with_odm.yaml:15]: Can you add a comment after the ratio that this will be overloaded as a reward dataset too in odm.

File: tests/test_sft_trainer.py [hunk 3]
```diff
@@ -2183,3 +2185,201 @@ def test_handler(element, **kwargs):
...
+        assert len(output_inference) > 0
+        assert (
+            "It takes 10 days for digging a trench of 100 m long, 50 m broad and 10 m deep. "
+            "What length of trench,\n25 m broad and 15 m deep can be dug in 30 days ?"
+            in output_inference
+        ), f"{output_inference} does not include the prompt"
+
+
+@pytest.mark.skipif(
+    not is_fms_accelerate_available(plugins="odm"),
```
Comment (dushyantbehl) [inline tests/test_sft_trainer.py:2309]: Are we enabling installing `odm` plugin by default in the `tox.ini` file?
if not i am assmuming these tests are run separately

File: tuning/data/data_processors.py [hunk 4]
```diff
@@ -452,9 +452,35 @@ def split_dataset(
...
+        train_split = "train"
+        eval_split = "test"
+        train_datasets_dict = {}
+        eval_datasets_dict = {}
+        for d, raw in processed_datasets:
+            if train_split in raw:
+                train_datasets_dict[d.name] = raw[train_split]
+            if eval_split in raw:
+                eval_datasets_dict[d.name] = raw[eval_split]
+        return train_datasets_dict, eval_datasets_dict
```
Comment (dushyantbehl) [inline tuning/data/data_processors.py:473]: Are we just looking to filter `train` and `test` splits from the dataset here?
Do you need the returned values to be a raw dict instead of a container?

File: tuning/data/setup_dataprocessor.py [hunk 5]
```diff
@@ -588,6 +595,10 @@ def process_dataargs(
             "Check your data config or ensure split sizes are valid."
         )
     if data_args.do_dataprocessing_only:
+        if odm_config:
+            raise ValueError(
+                "data processing with online data mixing is not currently supported"
```
Comment (dushyantbehl) [inline tuning/data/setup_dataprocessor.py:600]: Any reason we do not support data processing only mode for ODM config?

We can just dump the processed datasets and return right?

ODM then has to applied while training too but is there a fundamental problem in compatibility?

File: tuning/data/setup_dataprocessor.py [hunk 6]
```diff
@@ -613,20 +624,80 @@ def process_dataargs(
     # Note: This check should not be removed.
     #       Its important to recompute this post handling to
     #       check if we already tokenized the dataset or not.
-    is_tokenized_dataset = is_pretokenized_dataset(train_dataset or eval_dataset)
+    if odm_config:
+        is_tokenized_dataset = True
```
Comment (dushyantbehl) [inline tuning/data/setup_dataprocessor.py:628]: why are ODM datasets always pretokenized?

File: tuning/data/setup_dataprocessor.py [hunk 7]
```diff
@@ -613,20 +624,80 @@ def process_dataargs(
...
+    else:
+        is_tokenized_dataset = is_pretokenized_dataset(train_dataset or eval_dataset)
+
+    data_collator = None
+    if odm_config:
```
Comment (dushyantbehl) [inline tuning/data/setup_dataprocessor.py:633]: is it possible to wrap all of ODM stuff in a separate function and call it once inside?

File: tuning/data/setup_dataprocessor.py [hunk 8]
```diff
@@ -613,20 +624,80 @@ def process_dataargs(
...
+            max_seq_length,
+            data_args.instruction_template,
+            is_padding_free=is_padding_free,
+            processor=processor,
+        )
+    dataset_kwargs = {}
+    if odm_config:
+        # Third Party
+        # pylint: disable=import-outside-toplevel
+        if not is_fms_accelerate_available(plugins="odm"):
```
Comment (dushyantbehl) [inline tuning/data/setup_dataprocessor.py:676]: Shouldn't we do this check at the top as first thing?

File: tuning/data/setup_dataprocessor.py [hunk 9]
```diff
@@ -613,20 +624,80 @@ def process_dataargs(
...
+            None,
+            gamma=odm_config.odm.gamma,
+            eta=odm_config.odm.eta,
+            output_dir=train_args.output_dir,
+            sampling_interval=odm_config.odm.sampling_interval,
+            eval_batch_size=train_args.per_device_eval_batch_size,
+            reward_type=odm_config.odm.reward_type,
+        )
+        dataset_kwargs["skip_prepare_dataset"] = True
+        train_args.accelerator_config = {"split_batches": True}
```
Comment (dushyantbehl) [inline tuning/data/setup_dataprocessor.py:699]: Is `split_batches` needed for ODM?

File: tuning/sft_trainer.py [hunk 10]
```diff
@@ -125,6 +128,13 @@ def train(
     logger, train_args.log_level = set_log_level(
         logger_name="sft_trainer_train", level=train_args.log_level
     )
+
+    odm_config = None
+    if data_args.data_config_path:
```
Comment (dushyantbehl) [inline tuning/sft_trainer.py:133]: Can we please do this inside the `process_dataargs` function?

We want to keep `sft_trainer` clean from any data related functionality.

**Review by dushyantbehl — CHANGES_REQUESTED** [review]
Requesting minor revision of the documentation and restructing of code.

File: docs/advanced-data-preprocessing.md [hunk 11]
```diff
@@ -192,6 +203,59 @@ We also allow users to pass a [`seed`](https://huggingface.co/docs/datasets/v3.2
 
 Note: If a user specifies data sampling they can expect the datasets to be mixed and individual samples in the dataset to not be broken unless the max_seq_len argument is smaller than the length of individual samples in the dataset
 
+### Online Data Mixing
```
[the remark this answers is not in the document]
Comment (kmehant) [reply docs/advanced-data-preprocessing.md]: Done, I have made it into a new doc and changed references accordingly.

File: tests/test_sft_trainer.py (hunk 3 above)
Comment (kmehant) [reply tests/test_sft_trainer.py:2309]: No we dont want to install the odm plugin by default. Yes I have run these tests on my end:

Check the screenshots below:

Screenshot 2025-10-07 at 5 48 48 PM

Screenshot 2025-10-07 at 5 50 55 PM

File: tuning/data/data_processors.py (hunk 4 above)
Comment (kmehant) [reply tuning/data/data_processors.py:473]: > Are we just looking to filter train and test splits from the dataset here?

Yes 

> Do you need the returned values to be a raw dict instead of a container?

dict would be good over datasetdict but both ways it should work.

File: tuning/data/setup_dataprocessor.py (hunk 8 above)
Comment (kmehant) [reply tuning/data/setup_dataprocessor.py]: Can you mention where exactly?

File: tuning/data/data_processors.py [hunk 12]
```diff
@@ -504,7 +530,8 @@ def _process_dataset_configs(
 
             # Append the processed datasets to the final dict
             processed_datasets.append((d, raw_datasets))
-
+        if odm_config:
+            return self._process_datasets_for_odm(processed_datasets)
```
Comment (dushyantbehl) [inline tuning/data/data_processors.py:534]: I missed this last time but is sampling and concatenation of datasets not compatible with ODM?

Can we let the normal data processing perform its function i.e give a processed dataset and then wrap ODM on top...

The way I see is you don't need to modify code in `data_processors.py` the function `_process_datasets_for_odm` gets called after the datasets are returned by `_process_dataset_configs` and then you apply odm processing by calling this function inside `setup_dataprocessor.py`

File: tuning/data/setup_dataprocessor.py [hunk 13]
```diff
@@ -491,6 +495,72 @@ def dump_dataset(
         raise RuntimeError(f"Failed to dump dataset due to error {e}") from e
 
 
+def process_dataargs_odm(
```
Comment (dushyantbehl) [inline tuning/data/setup_dataprocessor.py:498]: nit: could we call this `setup_train_dataset_for_odm` 

because you are not touching eval_dataset

File: tuning/data/setup_dataprocessor.py [hunk 14]
```diff
@@ -491,6 +495,72 @@ def dump_dataset(
...
+    data_args: DataArguments,
+    tokenizer: AutoTokenizer,
+    train_args: TrainingArguments,
+    is_padding_free: bool = False,
+    processor: AutoProcessor = None,
+    odm_config: ODMConfig = None,
+    train_dataset: Dict = None,
+    eval_dataset: Dict = None,
```
Comment (dushyantbehl) [inline tuning/data/setup_dataprocessor.py:506]: can we rename this to `reward_dataset` and pass `eval_dataset` here after mentioning a comment in the code

File: tuning/data/setup_dataprocessor.py
Comment (dushyantbehl) [inline tuning/data/setup_dataprocessor.py:630]: We can potentially return `odm_config` from this function...or you can return the original `data_config` from this function to `odm_config` from inside it here.

We can initialize Acceleration framework ODM here too i think...this would save us loading data config twice

File: tuning/sft_trainer.py (hunk 10 above)
Comment (dushyantbehl) [reply tuning/sft_trainer.py:133]: @kmehant
can process_dataargs not initialize ODM framework and then return anything if needed...as far as I see the only thing we return is a dataset of type `ODM` so why can't we do the ODM framework initialization inside `process_dataargs` and return train and eval datasets as usual just `ODM` this time.

**Review by dushyantbehl — CHANGES_REQUESTED** [review]
Thanks for fixing the earlier comments...requesting some more clarifications.

File: tuning/data/setup_dataprocessor.py [hunk 15]
```diff
@@ -491,6 +495,72 @@ def dump_dataset(
...
+            tokenizer,
+            is_tokenized_dataset,
+            max_seq_length,
+            data_args.instruction_template,
+            is_padding_free=is_padding_free,
+            processor=processor,
+        )
+
+    # pylint: disable=import-outside-toplevel
+    if not is_fms_accelerate_available(plugins="odm"):
```
[the remark this answers is not in the document]
Comment (kmehant) [reply tuning/data/setup_dataprocessor.py:538]: done

File: docs/tuning-techniques.md [hunk 16]
```diff
@@ -470,6 +470,7 @@ The list of configurations for various `fms_acceleration` plugins:
   - `--multipack`: technique for *multi-gpu training* to balance out number of tokens processed in each device, to minimize waiting time.
 - [fast_moe_config](./tuning/config/acceleration_configs/fast_moe.py) (experimental):
   - `--fast_moe`: trains MoE models in parallel with [Scatter MoE kernels](https://github.com/foundation-model-stack/fms-acceleration/tree/main/plugins/accelerated-moe#fms-acceleration-for-mixture-of-experts), increasing throughput and decreasing memory usage.
+- [odm_config](./tuning/config/acceleration_configs/odm.py) (experimental): See [advanced data preprocessing](./advanced-data-preprocessing.md#online-data-mixing) for usage with data_config. This plugin allows dynamically mixing datasets online during training adapting to training signals.
```
[the remark this answers is not in the document]
Comment (kmehant) [reply docs/tuning-techniques.md:473]: Done!

File: tests/artifacts/predefined_data_configs/multiple_datasets_with_odm.yaml [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+      update_interval: 1 # update every step
+      sampling_interval: 1 # sample category for every sample
+      reward_type: validation_loss # uses eval loss of each dataset as reward
+      gamma: 0.1 # MAB hyper-parameter
+      eta: 0.2 # MAB hyper-parameter
+datasets:
+  - name: dataset_1
+    split:
+      train: 0.8
+      validation: 0.2 # validation set is also used in reward computation when reward_type is validation_loss.
```
[the remark this answers is not in the document]
Comment (kmehant) [reply tests/artifacts/predefined_data_configs/multiple_datasets_with_odm.yaml:15]: done!

File: tests/artifacts/predefined_data_configs/multiple_datasets_with_odm.yaml [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+        arguments:
+          remove_columns: all
+          batched: false
+          fn_kwargs:
+            input_column_name: input
+            output_column_name: output
+  - name: dataset_2
+    split:
+      train: 0.6
+      validation: 0.2 # validation set is also used in reward computation when reward_type is validation_loss.
```
[the remark this answers is not in the document]
Comment (kmehant) [reply tests/artifacts/predefined_data_configs/multiple_datasets_with_odm.yaml:30]: done!

File: tests/artifacts/predefined_data_configs/multiple_datasets_with_odm.yaml (hunk 17 above)
[the remark this answers is not in the document]
Comment (kmehant) [reply tests/artifacts/predefined_data_configs/multiple_datasets_with_odm.yaml:15]: There were couple more places so modified in all the files

File: tuning/data/setup_dataprocessor.py (hunk 13 above)
Comment (kmehant) [reply tuning/data/setup_dataprocessor.py:498]: True, done!

**Comment by kmehant** [comment]
Screenshot 2025-10-07 at 10 53 37 PM

tests on latest changes pass
